index.bs

```diff
@@ -3552,73 +3552,49 @@ Methods</h4>
-
-		<div algorithm="cancel and hold">
-			Let \(t_c\) be the value of <code>cancelTime</code>. Then
-
-			1. Let \(E_1\) be the event (if any) at time \(t_1\) where
-				\(t_1\) is the largest number satisfying \(t_1 \le t_c\).
-
-			2. Let \(E_2\) be the event (if any) at time \(t_2\) where
-				\(t_2\) is the smallest number satisfying \(t_c \lt t_2\).
-
-			3. If \(E_2\) exists:
-				1. If \(E_2\) is a linear or exponential ramp,
-					1. Effectively rewrite \(E_2\) to be the same kind of
-						ramp ending at time \(t_c\) with an end value that
-						would be the value of the original ramp at time
-						\(t_c\).
-						<img alt="Graphical representation of calling cancelAndHoldAtTime when linearRampToValueAtTime has been called at this time." src="images/cancel-linear.svg" style="height:75%;width:75%">
-
-					2. Go to step 5.
-
-				2. Otherwise, go to step 4.
-
-			4. If \(E_1\) exists:
-				1. If \(E_1\) is a <code>setTarget</code> event,
-					1. Implicitly insert a <code>setValueAtTime</code>
-						event at time \(t_c\) with the value that the
-						<code>setTarget</code> would have at time
-						\(t_c\).
-						<img alt="Graphical representation of calling cancelAndHoldAtTime when setTargetAtTime has been called at this time" src="images/cancel-setTarget.svg" style="height:75%;width:75%">
-
-					2. Go to step 5.
-
-				2. If \(E_1\) is a <code>setValueCurve</code> with a start
-					time of \(t_3\) and a duration of \(d\)
-
-					1. If \(t_c \gt t_3 + d\), go to step 5.
-
-					2. Otherwise,
-						1. Effectively replace this event with a
-							<code>setValueCurve</code> event with a start time
-							of \(t_3\) and a new duration of \(t_c-t_3\).
-							However, this is not a true replacement; this
-							automation MUST take care to produce the same
-							output as the original, and not one computed using
-							a different duration. (That would cause sampling of
-							the value curve in a slightly different way,
-							producing different results.)
-							<img alt="Graphical representation of calling cancelAndHoldAtTime when setValueCurve has been called at this time" src="images/cancel-setValueCurve.svg" style="height:75%;width:75%">
-
-						2. Go to step 5.
-
-			5. Remove all events with time greater than \(t_c\).
-
-			If no events are added, then the automation value after
-			{{AudioParam/cancelAndHoldAtTime()}} is the constant value that
-			the original timeline would have had at time \(t_c\).
-		</div>
+		When <code>cancelAndHoldAtTime</code> is called with a time <var>t</var> as
+		a parameter, execute the following steps:
+
+		- Let <var>v</var> be the value of the {{AudioParam}} at time
+			<var>t</var>.
+		- Let <var>current</var> be the <a lt="current audioparam">current
+			AudioParam</a> at time <var>t</var>.
+		- <a lt="rewrite event">Rewrite the current AudioParam event</a> to
+			end at <var>t</var>.
+		- Remove all {{AudioParam}} event that are in the list after
```

Comment:
"all automation events"

Reply:
> There is the issue of "being in the list after something". The spec mentions a "[list of automation events](https://webaudio.github.io/web-audio-api/#automation-event-time)". It says "The list of automation events is maintained in ascending order of automation event time.". But this is ambiguous. What is the "event time" in the case of a ramp (which has an [implicit] start time and an end time)? What if you decide to add an explicit `linearRamp(startTime, endTime, value)` variant (in fact, I'm planning to suggest such a change)?

The spec is frozen, we'll not be adding new features. Also it's very simply to do it in js and is kind of against the current model. For example `setValueCurveAtTime([startValue, endValue], startTime, duration);` is what you describe, but there are multiple ways of doing it.

Event have a time, it's either their start time or their end time. You're supposed to use another even to specify the other part, there is no implicit start time or anything. It's weird, but that's the way it is, and this ship has sailed long ago.

What's even more weird is that most event don't have a duration, but the curves have a duration. I'd have preferred a more unified model but that's life.

Reply:
> The spec is frozen, we'll not be adding new features.

Does that mean there's no point in asking for new features on GitHub? If so, perhaps you guys should make that clear – currently there's explicit "Feature request" UI when you post a new issue.

> Also it's very simply to do it in js and is kind of against the current model. For example `setValueCurveAtTime([startValue, endValue], startTime, duration);` is what you describe, but there are multiple ways of doing it.

I was actually thinking that there is no model to speak of, more like a counterintuitive design with different automation methods accepting different arguments.
Thanks for bringing setValueCurveAtTime() to my attention. Indeed it would seem to do what I want!

> > Event have a time, it's either their start time or their end time. You're supposed to use another even to specify the other part, there is no implicit start time or anything.

Yeah, I know how it currently works. I called the start time "implicit" because you don't specify it explictly when you call the function – it depends on the current state of the automation timeline (the time of the last event).

> What's even more weird is that most event don't have a duration, but the curves have a duration. I'd have preferred a more unified model but that's life.

I agree, that's another inconsistency! Who decides what can be changed or added to the spec? (Sorry, not familiar with how the W3C works.)

Reply:
> Event have a time, it's either their start time or their end time. You're supposed to use another even to specify the other part, there is no implicit start time or anything. It's weird, but that's the way it is, and this ship has sailed long ago.

Correct me if I'm wrong, but the spec does not define "event time" anywhere. [This](https://webaudio.github.io/web-audio-api/#automation-event-time) is not a definition. Perhaps you could fix it? (Or should I file a bug?)

Reply:
> Does that mean there's no point in asking for new features on GitHub? If so, perhaps you guys should make that clear – currently there's explicit "Feature request" UI when you post a new issue.

That is not what @padenot meant. We (the working group) are always open to feature request. But the spec document goes through the W3C process and it is at the final stage of such process, so any significant change will be addressed in V2 work or other processes.

> I agree, that's another inconsistency!

As you can see, @padenot and myself agree that this is not ideal. But also the AudioParam interface has more than 7 years of history. We cannot simply change that because now we realized it is a bad thing - that change will break almost everything built with Web Audio API.

> the spec does not define "event time" anywhere

The term "event time" is so generic, so the description at the link on you provided seems better. So we might want to use "automation event time" where it's possible.

Reply:
@hoch OK, so I understand I can still make feature requests for V2 using the GitHub form.
I understand the need for maintaining compatibility and I am not proposing any breaking changes – rather adding alternative variants with more sensible conventions.

> The term "event time" is so generic, so the description at the link on you provided seems better

Not sure what you mean here, so I'll just try to make my point again: The term "event time" as used in the spec is quite counterintuitive -- for some automations it's the start time, for others it's the end time. That's why it needs a definition. I personally had serious doubts every time I read about "event time" in the spec (eventually I figured out that it's probably whatever time is passed into the function, but I was never 100% sure). I'm probably not the only one.

Reply:
I think we just need to define "automation event time" a bit better.  Every event today has exactly one time parameter.  This is the event time.